pkg/api/kptfile/v1alpha2/types.go

```diff
@@ -222,6 +222,12 @@ func (p *Pipeline) IsEmpty() bool {
 
 // Function specifies a KRM function.
 type Function struct {
+	// `Name` specifies the name of the function within the pipeline. It must
+	// be unique within the pipeline.
+	// The name will be used as the associative key when merging changes
+	// from the upstream package into the pipeline.
+	Name string `yaml:"name,omitempty"`
```

Comment:
I like this idea. But is this required field ? What if user doesn't specify this ? Should we error out ?

Reply:
Meta commentary: Schema changes should be done separately, not buried in implementation PRs.

What's the "name of the function" besides the Image name. I'm guessing you mean name of this particular invocation? If it's only required for merging, it will seem extraneous to most users. I suggest starting with the design of the merge behavior and it if it makes sense to the reader.

Reply:
I think the name here can best be thought of as the identifier for the instance of a function. It would be a required field, so it would be an error if it isn't specified.

In order to a 3-way merge lists of objects (associative arrays), we need a key to be able to match elements from upstream, local and origin. Kubernetes usually handles this with a `name` property and amending the openapi schema with information about which property should be the key (Using the `x-kubernetes-patch-merge-key` openapi extension). Sometimes the associative key is named something else than `name`.

We could use the `image` value as the key (and exclude the version tag to avoid some conflicts), but I think using the 
same function multiple times in the same pipeline is a valid use-case. Trying to merge without a key will not work well.

Reply:
I will break this change out into a separate PR if/when this looks good otherwise.

Reply:
1. Let's split the schema change to a separate PR as it has far reaching implications. In that PR, start with just the schema change, and once we're all in agreement then proceed to change a) Validation logic of Pipeline b) Existing tests c) Examples (it's also fine to delegate this work by filing an issue).
2. One difference between this and "associated arrays" like PodSpec.containers is that the order is actually important and has a defined semantic. It's a real list, not a key/value mapping encoded as a list of objects with a designated key. Is there an example like this? The whole merging strategy has been a mine field both in api machinary and kustomize, so I like to better understand the latest and greatest thinking on how to formulate this use case.

Reply:
So the goal of merging associative arrays like PodSpec.containers and this is the same, although I agree the consequences of getting it wrong is worse in this case. The challenge with merging of ports in kustomize is a [complex issue](https://github.com/kubernetes/kubernetes/issues/39188) and not really related to the merge logic itself, but it does show the challenge of not having a good merge key.

Not sure I follow the question around `{"key": value}`. Mapping nodes in yaml are defined as an unordered set of keys. But for the merge logic, we essentially just need a property (or multiple properties together) that uniquely identifies each element in the list. I don't think SSA "solves" merging of lists and as far as I know it still requires elements to have unique associative keys, but I will verify.

I agree we can decide ourselves on the rules for merge here, but it is hard to come up with a sensible strategy for doing the merge if we don't have a key. Admittedly, it is hard even when we have one.

Reply:
If it is required field, it's good to have it as a property of mutators. 

```
pipeline:
  mutators:
    ApplySetters1:
        image: gcr.io/kpt-fn/apply-setters:unstable
        configMap:
          foo: bar
    ApplySetters2:
        image: gcr.io/kpt-fn/apply-setters:unstable
        configMap:
          baz: bor
```

This way, the schema can be extended to provide additional constraints on the inputs.